chains/solana/contracts/programs/ccip-offramp/src/instructions/v1/execute.rs

```diff
@@ -368,104 +418,194 @@ fn internal_execute<'info>(
+    }
+
+    pub fn get_token_accounts_for<'b>(
+        &'b self,
+        router: Pubkey,
+        fee_quoter: Pubkey,
+        token_receiver: Pubkey,
+        chain_selector: u64,
+        i: usize,
+    ) -> Result<TokenAccounts<'a>> {
+        let accs = self
+            .token_accounts_per_token
+            .get(i)
+            .map(|accounts_per_token| accounts_per_token.accounts)
+            .ok_or(CcipOfframpError::InvalidInputsTokenIndices)?;
+
+        validate_and_parse_token_accounts(
+            token_receiver,
+            chain_selector,
+            router,
+            fee_quoter,
+            Some(crate::ID),
+            accs,
+        )
     }
 
-    Ok(MessagingAccounts {
-        logic_receiver,
-        external_execution_signer,
-        external_execution_signer_bump: signer_bump,
-        remaining_accounts: msg_accounts,
-    })
+    pub fn new(
+        remaining_accounts: &'a [AccountInfo<'a>],
+        token_indices: &[u8],
+        report_is_buffered: bool,
+        source_bitmap: &u64,
+    ) -> Result<Self> {
+        ////////////////////////
+        // Buffering subslice //
+        ////////////////////////
+
+        // First, if the report is buffered, it means the last account is the buffering account.
+        // We can simply remove it from consideration from now on.
+        let (remaining_accounts_without_buffering, buffering_accounts) = if report_is_buffered {
+            (
+                &remaining_accounts[0..(remaining_accounts.len() - 1)],
+                Some(ExecuteContextRemainingBufferingAccounts {
+                    execution_buffer_account: remaining_accounts
+                        .last()
+                        .ok_or(CcipOfframpError::ExecutionReportUnavailable)?,
+                }),
+            )
+        } else {
+            (remaining_accounts, None)
+        };
+
+        ////////////////////////
+        // Messaging subslice //
+        ////////////////////////
+        let only_tokens = token_indices.first().map(|i| *i == 0).unwrap_or_default();
+        let messaging_accounts_exist =
+            !remaining_accounts_without_buffering.is_empty() && !only_tokens;
+        let messaging_accounts = if messaging_accounts_exist {
+            let end_index = token_indices
+                .first()
+                .map(|i| *i as usize)
+                .unwrap_or(remaining_accounts_without_buffering.len());
+
+            require!(
+                1 <= end_index && end_index <= remaining_accounts_without_buffering.len(), // program id and message accounts need to fit in remaining accounts
+                CcipOfframpError::InvalidInputsTokenIndices
+            ); // there could be other remaining accounts used for tokens
+
+            let logic_receiver = &remaining_accounts_without_buffering[0];
+            let external_execution_signer = &remaining_accounts_without_buffering[1];
+            let msg_accounts = &remaining_accounts_without_buffering[2..end_index];
```

Comment:
Unrelated to your changes, just a comment to think about and maybe do a follow-up PR on:

There's something slightly off here with regards to the `end_index`. We validate that it is `>=1`, and then do `let msg_accounts = &remaining_accounts_without_buffering[2..end_index];` which doesn't make much sense when the `end_index` is `=1`. It looks like the initial restriction might be slightly too permissive, and the actual lowest admissible `end_index` should be 2 or 3 here.